### ISSUE TITLE
`SemijoinIdempotence` doesn't kick in when the join condition has additional non-column-equality predicates

### ISSUE THREAD

**Issue by ggevay** [MaterializeInc/materialize#20632]
### What version of Materialize are you using?

8bd302f369a6a902679e34d68398ebb45a0c506a

### What is the issue?

_Note: This issue should probably be broken into two issues based on whether the extra join condition concerns the left or the right input. [The right input case looks significantly easier.](https://github.com/MaterializeInc/materialize/issues/20632#issuecomment-1651659601)_

[This is from a customer query.](https://materializeinc.slack.com/archives/C04LQV4KUR4/p1689678531450239?thread_ts=1689610328.454189&cid=C04LQV4KUR4) Simplified query to demonstrate the problem:

```
materialize=> 
create table t1(x int, y int, t int);
insert into t1 values (1, 10, -1), (2, 20, -1), (3, 30, -2);

create table t2(x int, y int);
insert into t2 values (1, 10), (2, 20), (3, 30);

create view t2_keyed as select distinct on(x) * from t2;

CREATE TABLE
INSERT 0 3
CREATE TABLE
INSERT 0 3
CREATE VIEW
materialize=> 

-- SemijoinIdempotence kicks in (only one Join in the optimized plan)
explain
select * from
t1 left join t2_keyed on t1.x = t2_keyed.x;
               Optimized Plan                
---------------------------------------------
 Explained Query:                           +
   Return                                   +
     Union                                  +
       Map (null, null)                     +
         Union                              +
           Negate                           +
             Project (#0..=#2)              +
               Get l0                       +
           Get materialize.public.t1        +
       Project (#0..=#2, #0, #3)            +
         Get l0                             +
   With                                     +
     cte l0 =                               +
       Project (#0..=#2, #4)                +
         Join on=(#0 = #3) type=differential+
           ArrangeBy keys=[[#0]]            +
             Filter (#0) IS NOT NULL        +
               Get materialize.public.t1    +
           ArrangeBy keys=[[#0]]            +
             TopK group_by=[#0] limit=1     +
               Filter (#0) IS NOT NULL      +
                 Get materialize.public.t2  +
                                            +
 Source materialize.public.t2               +
   filter=((#0) IS NOT NULL)                +
 
(1 row)

materialize=> 

-- SemijoinIdempotence doesn't kick in (two Joins in the optimized plan)
explain
select * from
t1 left join t2_keyed on t1.x = t2_keyed.x and t1.t = -1;
                            Optimized Plan                             
-----------------------------------------------------------------------
 Explained Query:                                                     +
   Return                                                             +
     Union                                                            +
       Project (#0..=#2, #0, #3)                                      +
         Get l0                                                       +
       Project (#0..=#2, #6, #7)                                      +
         Map (null, null)                                             +
           Join on=(#0 = #3 AND #1 = #4 AND #2 = #5) type=differential+
             ArrangeBy keys=[[#0..=#2]]                               +
               Union                                                  +
                 Map (-1)                                             +
                   Negate                                             +
                     Distinct group_by=[#0, #1]                       +
                       Project (#0, #1)                               +
                         Get l0                                       +
                 Distinct group_by=[#0..=#2]                          +
                   Get materialize.public.t1                          +
             ArrangeBy keys=[[#0..=#2]]                               +
               Get materialize.public.t1                              +
   With                                                               +
     cte l0 =                                                         +
       Project (#0..=#2, #4)                                          +
         Join on=(#0 = #3) type=differential                          +
           ArrangeBy keys=[[#0]]                                      +
             Filter (#2 = -1) AND (#0) IS NOT NULL                    +
               Get materialize.public.t1                              +
           ArrangeBy keys=[[#0]]                                      +
             TopK group_by=[#0] limit=1                               +
               Filter (#0) IS NOT NULL                                +
                 Get materialize.public.t2                            +
                                                                      +
 Source materialize.public.t2                                         +
   filter=((#0) IS NOT NULL)                                          +
 
(1 row)
```

From looking at the traces, this is how the plan looks like just before the successful SemijoinIdempotence:
```
optimize/local/logical/fixpoint
Return // { arity: 5 }
  Union // { arity: 5 }
    Map (null, null) // { arity: 5 }
      Union // { arity: 3 }
        Project (#0..=#2) // { arity: 3 }
          Negate // { arity: 4 }
            Join on=(#0 = #3) // { arity: 4 }
              Filter (#2 = -1) // { arity: 3 }
                Get materialize.public.t1 // { arity: 3 }
              Distinct group_by=[#0] // { arity: 1 }
                Get l0 // { arity: 5 }
        Filter (#2 = -1) // { arity: 3 }
          Get materialize.public.t1 // { arity: 3 }
    Project (#0..=#2, #0, #4) // { arity: 5 }
      Filter (#2 = -1) // { arity: 5 }
        Get l0 // { arity: 5 }
With
  cte l0 =
    Join on=(#0 = #3) // { arity: 5 }
      Filter (#0) IS NOT NULL // { arity: 3 }
        Get materialize.public.t1 // { arity: 3 }
      Filter (#0) IS NOT NULL // { arity: 2 }
        Get materialize.public.t2_keyed // { arity: 2 }
```
And this is the plan before the unsuccessful SemijoinIdempotence:
```
optimize/local/logical/fixpoint
Return // { arity: 5 }
  Union // { arity: 5 }
    Project (#0..=#2, #0, #4) // { arity: 5 }
      Get l0 // { arity: 5 }
    Project (#0..=#2, #6, #7) // { arity: 5 }
      Join on=(#0 = #3 AND #1 = #4 AND #2 = #5) // { arity: 8 }
        Union // { arity: 3 }
          Negate // { arity: 3 }
            Distinct group_by=[#0, #1, -1] // { arity: 3 }
              Get l0 // { arity: 5 }
          Distinct group_by=[#0..=#2] // { arity: 3 }
            Get materialize.public.t1 // { arity: 3 }
        Get materialize.public.t1 // { arity: 3 }
        Constant // { arity: 2 }
          - (null, null)
With
  cte l0 =
    Join on=(#0 = #3) // { arity: 5 }
      Filter (#0) IS NOT NULL AND (#2 = -1) // { arity: 3 }
        Get materialize.public.t1 // { arity: 3 }
      Filter (#0) IS NOT NULL // { arity: 2 }
        Get materialize.public.t2_keyed // { arity: 2 }
```
We are suspecting that it's the 
```
        Constant
          - (null, null)
```
that is preventing the SemijoinIdempotence. @aalexandrov observed that this probably comes from [the "more general" outer join lowering branch](https://github.com/MaterializeInc/materialize/blob/fa1a8893b3ce27a29cd648e3dd84e92f366c1829/src/sql/src/plan/lowering.rs#L597). (See [here](https://github.com/MaterializeInc/materialize/blob/fa1a8893b3ce27a29cd648e3dd84e92f366c1829/src/sql/src/plan/lowering.rs#L1867).)

Note that the above Constant join input is later eliminated by a combination of LiteralLifting and then RedundantJoin, after which the fixpoint loop runs SemijoinIdempotence again. However, SemijoinIdempotence still has no luck, because by that time the plan looks like this:
```
optimize/local/logical/fixpoint/0000
Return // { arity: 5 }
  Union // { arity: 5 }
    Project (#0..=#2, #0, #4) // { arity: 5 }
      Get l0 // { arity: 5 }
    Project (#0..=#2, #6, #7) // { arity: 5 }
      Map (null, null) // { arity: 8 }
        Join on=(#0 = #3 AND #1 = #4 AND #2 = #5) // { arity: 6 }
          Union // { arity: 3 }
            Map (-1) // { arity: 3 }
              Negate // { arity: 2 }
                Distinct group_by=[#0, #1] // { arity: 2 }
                  Get l0 // { arity: 5 }
            Distinct group_by=[#0..=#2] // { arity: 3 }
              Get materialize.public.t1 // { arity: 3 }
          Get materialize.public.t1 // { arity: 3 }
With
  cte l0 =
    Join on=(#0 = #3) // { arity: 5 }
      Filter (#0) IS NOT NULL AND (#2 = -1) // { arity: 3 }
        Get materialize.public.t1 // { arity: 3 }
      Filter (#0) IS NOT NULL // { arity: 2 }
        Get materialize.public.t2_keyed // { arity: 2 }
```

I think the problem in this run is that in the meantime LiteralLifting created the `Map (-1)`.

I can see several ways to fix this, but none of them are particularly simple:
- Smarten SemijoinIdempotence to be able to run either on the first version or the second version. (I don't know where exactly the pattern matching fails currently, so I'm not sure how hard this would be. Certainly not very easy, because SemijoinIdempotence is super complicated.)
- Smarten the outer join lowering to already do what LiteralLifting would do later to eliminate that constant `(null, null)` input.
- Perform a modified version of LiteralLifting and RedundantJoin already before SemijoinIdempotence. This run of LiteralLifting shouldn't lift stuff from Filters, only from Constants, so that it doesn't create the `Map (-1)`, but allows RedundantJoin to eliminate the constant `(null, null)`.

cc @aalexandrov

**Comment by aalexandrov** [comment]
We discussed this a bit further. The more general outer join planning pattern has a different `Join` shape for the "outer" part, so it's not immediately clear why and how `SemijoinIdempotence` should apply in that case.

**Comment by ggevay** [comment]
One more similar situation: https://materializeinc.slack.com/archives/C02PPB50ZHS/p1690297137310809?thread_ts=1690294151.692909&cid=C02PPB50ZHS
Here, the extra condition is on the right side of the LEFT JOIN, so it looks like an easier optimization. (It's also easy to do manually, but we still should do it, because the users won't think of doing this.)

**Comment by aalexandrov** [comment]
I hope that this can be closed for the same reasons as #20801 (should be resolved in #22343).

**Comment by ggevay** [comment]
Yes, the https://github.com/MaterializeInc/materialize/issues/22348 task in that epic should cover this.
